kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/KotlinSymbolProcessing.kt

```diff
@@ -8,66 +32,112 @@ import com.intellij.openapi.util.Disposer
-}
+class KotlinSymbolProcessing(
+    val compilerConfiguration: CompilerConfiguration,
+    val options: KspOptions,
+    val logger: KSPLogger,
+    val testProcessor: SymbolProcessorProvider? = null
+) {
 
-fun main(args: Array<String>) {
-    val kotlinSourceRoots = args.toList().map { File(it) }
-    val compilerConfiguration = CompilerConfiguration()
-    compilerConfiguration.put(CLIConfigurationKeys.MESSAGE_COLLECTOR_KEY, MessageCollector.NONE)
-    val env = KotlinCoreEnvironment.createForProduction(
+    val providers: List<SymbolProcessorProvider>
+    val env: KotlinCoreEnvironment = KotlinCoreEnvironment.createForProduction(
         Disposer.newDisposable(), compilerConfiguration,
         EnvironmentConfigFiles.JVM_CONFIG_FILES
     )
+    val project = env.project as MockProject
+    val kspCoreEnvironment = KSPCoreEnvironment(project)
 
-    val application = ApplicationManager.getApplication() as MockApplication
-    configureApplicationEnvironment(application)
+    var finished = false
+    val deferredSymbols = mutableMapOf<SymbolProcessor, List<KSAnnotated>>()
+    val ktFiles = convertFilesToKtFiles(project, compilerConfiguration.kotlinSourceRoots.map { it.path })
+    val codeGenerator: CodeGeneratorImpl
+    val processors: List<SymbolProcessor>
 
-    val files = kotlinSourceRoots
-        .sortedBy { Files.isSymbolicLink(it.toPath()) } // Get non-symbolic paths first
-        .flatMap { root -> root.walk().filter { it.isFile && it.extension == "kt" }.toList() }
-        .sortedBy { Files.isSymbolicLink(it.toPath()) } // This time is for .java files
-        .distinctBy { it.canonicalPath }
-    compilerConfiguration.addKotlinSourceRoots(files.map { it.absolutePath })
+    init {
+        configureProjectEnvironment(
+            project,
+            compilerConfiguration,
+            env::createPackagePartProvider,
+            env.projectEnvironment.environment.jarFileSystem as CoreJarFileSystem
+        )
 
-    val project = env.project as MockProject
-    val ktFiles = convertFilesToKtFiles(project, files)
-    configureProjectEnvironment(
-        project,
-        compilerConfiguration,
-        env::createPackagePartProvider,
-        env.projectEnvironment.environment.jarFileSystem as CoreJarFileSystem
-    )
-    val kspCoreEnvironment = KSPCoreEnvironment(project)
+        val ksFiles = ktFiles.map { KSFileImpl(it) }
+        val anyChangesWildcard = AnyChanges(options.projectBaseDir)
+        codeGenerator = CodeGeneratorImpl(
+            options.classOutputDir,
+            options.javaOutputDir,
+            options.kotlinOutputDir,
+            options.resourceOutputDir,
+            options.projectBaseDir,
+            anyChangesWildcard,
+            ksFiles,
+            options.incremental
+        )
+        val application = ApplicationManager.getApplication() as MockApplication
+        configureApplicationEnvironment(application)
+        providers = if (testProcessor != null) {
```

Comment:
Can we move this out of `KotlinSymbolProcessing` and pass the providers from outside?